Snowflake.Data/Core/SFChunkDownloader.cs

```diff
@@ -60,38 +57,63 @@ class SFChunkDownloader
             FillDownloads();
         }
 
-        private BlockingCollection<Task<SFResultChunk>> _downloadTasks;
-        
-        private void FillDownloads()
-        {
-            _downloadTasks = new BlockingCollection<Task<SFResultChunk>>(prefetchSlot);
+        private BlockingCollection<Lazy<Task<SFResultChunk>>> _downloadTasks;
+        private ConcurrentQueue<Lazy<Task<SFResultChunk>>> _downloadQueue;
 
-            Task.Run(() =>
+        private void RunDownloads()
+        {
+            try
             {
-                foreach (var c in chunks)
+                while (_downloadQueue.TryDequeue(out var task) && !externalCancellationToken.IsCancellationRequested)
                 {
-                    _downloadTasks.Add(DownloadChunkAsync(new DownloadContext()
+                    if (!task.IsValueCreated)
                     {
-                        chunk = c,
-                        chunkIndex = nextChunkToDownloadIndex,
-                        qrmk = this.qrmk,
-                        chunkHeaders = this.chunkHeaders,
-                        cancellationToken = this.externalCancellationToken
-                    }));
+                        task.Value.Wait(externalCancellationToken);
+                    }
                 }
+            }
+            catch (Exception)
+            {
+                //Don't blow from background threads.
+            }
+        }
+
+
+        private void FillDownloads()
+        {
+            _downloadTasks = new BlockingCollection<Lazy<Task<SFResultChunk>>>();
+
+            foreach (var c in chunks)
+            {
+                var t = new Lazy<Task<SFResultChunk>>(() => DownloadChunkAsync(new DownloadContext()

```

Comment:
Does `Lazy` means that Task will be executed only when adding task to collections>?

Reply:
Lazy means that the Task won't be executed until the `Value` property is called.